crates/core/src/factories/core_fetch.rs

```diff
@@ -379,19 +385,25 @@ impl CoreFetch {
                     ?peer_url,
                     "could not send fetch request: {err}."
                 );
+
                 // Remove all requests to that peer and notify if drained.
-                let mut lock = state.lock().expect("poisoned");
-                lock.requests.retain(|(_, a)| *a != peer_url);
-                Self::notify_listeners_if_queue_drained(lock);
+                {
+                    let mut lock = state.lock().expect("poisoned");
+                    lock.requests.retain(|(_, a)| *a != peer_url);
+                }
+                Self::notify_listeners_if_queue_drained(state.clone());
```

Comment:
This is not exactly the same, because there are multiple of these receivers processing outgoing fetch requests. It may be fine to change to this behavior, but it could also mean that we're seeing a change in gossip behavior in that it's not triggered as often during initial sync.
I'd prefer to keep the changes at a minimum while we're tying to find the issues.